### PR TITLE
Add diagrams and discussion of native driver execution type

### LINKED ISSUE #926: Doc bug: Add execution context changes to docs
### Describe the bug

Issue #794 has made changes to EC and API in load generated BPF driver. Please reflect this in the docs.

### Affected OS

- [ ] Windows Server 2022
- [ ] Windows Server 2019
- [ ] Windows 11
- [ ] Windows 10
- [ ] Windows Insider Preview (specify affected build below)
- [ ] Other (specify below)

### Additional OS information

_No response_

### Steps taken to reproduce bug

Read docs.
See that they are incorrect in that they do not reflect the changes made in issue #794

### Expected behavior

Docs should correctly reflect the behavior changes in issue #794

### Actual outcome

Docs are incorrect.

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by dthaler**
Signed-off-by: Dave Thaler <dthaler@microsoft.com>

## Description

You can see what the pages will look like at https://github.com/dthaler/ebpf-for-windows/blob/native/README.md

Fixes #926

## Testing

N/A

## Documentation

This entire PR is doc changes.

### CODE REVIEW

**Review by Alan-Jowett — DISMISSED** [review]

File: README.md [hunk 1]
```diff
@@ -98,4 +102,10 @@ maintained by the hypervisor.
...
 Since a hypervisor doing such code integrity checks will refuse to accept code pages that aren't signed by
 a key that the hypervisor trusts, this does impact eBPF programs running natively.  As such, when HVCI
-is enabled, eBPF programs work fine in interpreted mode, but not when using JIT compilation.
+is enabled, eBPF programs work fine in interpreted mode, but not when using JIT compilation because the JIT
+compiler does not have a key that the hypervisor trusts.  And since interpreted
+mode is absent in release builds, neither mode will work on an HVCI-enabled production system.
+
+Instead, a third mode is also supported by eBPF for Windows, in addition to JIT compiled and interpreted modes.
+This third mode entails compiling eBPF programs into regular Windows drivers that can be accepted by HVCI.
+For more discussion, see the [Native Code Generation documentation](docs/NativeCodeGeneration.md).
```
Comment (saxena-anurag) [inline README.md:111]: This link gives page not found error.

File: docs/NativeCodeGeneration.md [hunk 2]
```diff
@@ -23,9 +23,13 @@ image that can be loaded and used as an eBPF program.
 
 # Native Code Generation Pipeline
 
+The following diagram shows the steps in the pipeline:
+
+![eBPF Driver Toolchain](DriverToolchain.png)
```
Comment (saxena-anurag) [inline docs/NativeCodeGeneration.md:28]: This link also looks broken.

File: docs/NativeCodeGeneration.md [hunk 3]
```diff
@@ -73,13 +77,13 @@ driver skeleton.
...
 contracts exposed by the other framework components (to export the program to the eBPF runtime, to create any maps
 required, and to import the helper functions).
 
 ## Global entry point for generated code
 
-Every generated C file will contain a single global entry point of type metadata_table_t:
+Every generated C file contains a single global entry point of type metadata_table_t:
 
 ```
 typedef struct _metadata_table
```
Comment (saxena-anurag) [inline docs/NativeCodeGeneration.md:89]: This is obsolete now. Metadata table now looks like following:

` typedef struct _metadata_table
 {
        void (*programs)(program_entry_t** programs, size_t* count);
        void (*maps)(map_entry_t** maps, size_t* count);
    } metadata_table_t;`

File: docs/NativeCodeGeneration.md [hunk 4]
```diff
@@ -73,13 +77,13 @@ driver skeleton.
 
 All Windows Drivers need to export a static function labelled &quot;DriverEntry&quot; that is invoked by the OS
 (Operating System) during loading of the driver. In addition to the standard driver related functionality, the driver
-entry skeleton would also register three [Network Module Registrar (NMR)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/introduction-to-the-network-module-registrar)
+entry skeleton also registers three [Network Module Registrar (NMR)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/introduction-to-the-network-module-registrar)
```
Comment (saxena-anurag) [inline docs/NativeCodeGeneration.md:80]: The driver actually provides address of metadata table as part of client data when attach happens. Will that be 1 contract?

File: docs/NativeCodeGeneration.md [hunk 5]
```diff
@@ -109,12 +113,12 @@ typedef struct _program_entry
 } program_entry_t;
```
Comment (saxena-anurag) [inline docs/NativeCodeGeneration.md:113]: This is the updated `program_entry_t`

`typedef struct _program_entry
    {
        uint64_t (*function)(void*);
        const char* section_name;
        const char* program_name;
        uint16_t* referenced_map_indices;
        uint16_t referenced_map_count;
        helper_function_entry_t* helpers;
        uint16_t helper_count;
        size_t bpf_instruction_count;
        ebpf_program_type_t* program_type;
        ebpf_attach_type_t* expected_attach_type;
    } program_entry_t;`

File: README.md (hunk 1 above)
Comment (saxena-anurag) [reply README.md:111]: looks like this is not showing as this PR is not yet merged to main. Please ignore.

File: docs/NativeCodeGeneration.md (hunk 2 above)
Comment (saxena-anurag) [reply docs/NativeCodeGeneration.md:28]: looks like this is not showing as this PR is not yet merged to main. Please ignore.

File: docs/NativeCodeGeneration.md (hunk 5 above)
Comment (dthaler) [reply docs/NativeCodeGeneration.md:118]: Thanks, updated.

File: docs/NativeCodeGeneration.md (hunk 3 above)
Comment (dthaler) [reply docs/NativeCodeGeneration.md:89]: Thanks, updated.

File: docs/NativeCodeGeneration.md (hunk 4 above)
Comment (dthaler) [reply docs/NativeCodeGeneration.md]: Updated.

**Review by Alan-Jowett — APPROVED** [review]

**Review by saxena-anurag — APPROVED** [review]
Filed #984 to update `NativeCodeGeneration.md`

**Pull request merged**